src/helmRepoIndex/main.go

```diff
@@ -0,0 +1,135 @@
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/google/go-github/v48/github"
+	"golang.org/x/oauth2"
+	"helm.sh/helm/v3/pkg/repo"
+)
+
+func main() {
+	var centralHelmIndex string
+	var gitOwner string
+	var gitToken string
+
+	flag.StringVar(&centralHelmIndex, "indexFile", "", "Specify Helm Repository index file")
+	flag.StringVar(&gitOwner, "owner", "", "Specify GitHub User or Organization")
+	flag.StringVar(&gitToken, "token", "", "Specify GitHub Token")
+	flag.Parse()
+
+	ctx := context.Background()
+	client := getAuthenticatedClient(ctx, gitToken)
+	repos, _, _ := getOrgRepos(ctx, gitOwner, client)
+
+	for i, gitRepo := range repos {
+		var gitRepoHelmIndex string
+
+		//check gh pages configured for repo
+		_, response, _ := client.Repositories.GetBranch(ctx, gitOwner, *gitRepo.Name, "gh-pages", false)
+
+		if response.StatusCode == 200 {
+			fmt.Println(i, *gitRepo.Name)
+			gitRepoHelmIndex = downloadProductHelmRepoIndex(ctx, client, gitOwner, *gitRepo.Name)
+			buildHelmRepoIndex(centralHelmIndex, gitRepoHelmIndex)
+		}
+	}
+}
+
+func getAuthenticatedClient(ctx context.Context, gitToken string) *github.Client {
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: gitToken},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+
+	return github.NewClient(tc)
+}
+
+func getOrgRepos(ctx context.Context, gitOwner string, client *github.Client) ([]*github.Repository, *github.Response, error) {
+	opt := &github.RepositoryListByOrgOptions{
+		Type: "all",
+	}
+	repos, response, err := client.Repositories.ListByOrg(ctx, gitOwner, opt)
+	return repos, response, err
+}
+
+func downloadProductHelmRepoIndex(ctx context.Context, client *github.Client, gitOwner string, gitRepo string) string {
+	pageInfo, _, _ := client.Repositories.GetPagesInfo(ctx, gitOwner, gitRepo)
+	fullURLFile := *pageInfo.HTMLURL + "index.yaml"
+	fileName := gitRepo + "-index.yaml"
+
+	file, err := os.Create(fileName)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	response, _ := http.Get(fullURLFile)
+	if response.StatusCode == 200 {
+		_, _ = io.Copy(file, response.Body)
+
+		// Close response body
+		err := response.Body.Close()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		// Close file
+		err = file.Close()
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	return fileName
+}
+
+func buildHelmRepoIndex(indexFile string, mergeIndexFile string) {
```

Comment:
You are failing the whole build process if one repo produces an issue. we should make it more resiliant and skip a repository with a warning/error log if that happens

Reply:
Hm, why? func buildHelmRepoIndex is executed when repository has a valid GH pages setup but if GH Page delivers no `index.yaml` no merge will be executed.

Do you have another issue in mind? Ok, currently we rely on a consistent downloaded `index.yaml` if there is one.